Add support for python 3.12

I have a private repository that uses this action and when we tried to upgrade the python version to 3.12, we use it like this in our yaml file:
```yaml
      - name: Publish Unit Test Results
        uses: EnricoMi/publish-unit-test-result-action/linux@v2
        if: github.actor != 'dependabot[bot]'
        with:
          event_name: Integration Tests
          check_name: Integration Tests After Approval
          report_individual_runs: true
```
But after trying to upgrade the python version from 3.10.13 (on which it works well) to 3.12.0, this is the log:
```
> Install Python dependencies
...
  Collecting cffi==1.15.1 (from -r /home/ubuntu/.../_actions/EnricoMi/publish-unit-test-result-action/v2/linux/../python/requirements.txt (line 12))
    Using cached cffi-1.15.1.tar.gz (508 kB)
...
Building wheels for collected packages: cffi
    Building wheel for cffi (pyproject.toml): started
    Building wheel for cffi (pyproject.toml): finished with status 'error'
    error: subprocess-exited-with-error
    
    × Building wheel for cffi (pyproject.toml) did not run successfully.
    │ exit code: 1
    ╰─> [36 lines of output]
        running bdist_wheel
        running build
        running build_py
        creating build
        creating build/lib.linux-x86_64-cpython-312
        creating build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/__init__.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/pkgconfig.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/recompiler.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/vengine_cpy.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/model.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/backend_ctypes.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/api.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/setuptools_ext.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/commontypes.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/ffiplatform.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/error.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/lock.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/vengine_gen.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/cffi_opcode.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/verifier.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/cparser.py -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/_cffi_include.h -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/parse_c_type.h -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/_embedding.h -> build/lib.linux-x86_64-cpython-312/cffi
        copying cffi/_cffi_errors.h -> build/lib.linux-x86_64-cpython-312/cffi
        running build_ext
        building '_cffi_backend' extension
        creating build/temp.linux-x86_64-cpython-312
        creating build/temp.linux-x86_64-cpython-312/c
        gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DFFI_BUILDING=1 -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/include/ffi -I/usr/include/libffi -I/home/ubuntu/.../enricomi-publish-action-venv/include -I/opt/hostedtoolcache/Python/3.12.0/x64/include/python3.12 -c c/_cffi_backend.c -o build/temp.linux-x86_64-cpython-312/c/_cffi_backend.o
        c/_cffi_backend.c:15:10: fatal error: ffi.h: No such file or directory
           15 | #include <ffi.h>
              |          ^~~~~~~
        compilation terminated.
        error: command '/usr/bin/gcc' failed with exit code 1
        [end of output]
    
    note: This error originates from a subprocess, and is likely not a problem with pip.
    ERROR: Failed building wheel for cffi
  ERROR: Could not build wheels for cffi, which is required to install pyproject.toml-based projects
  Failed to build cffi
  Error: Process completed with exit code 1.
```

I guess this error makes sense since this action pins `cffi` to `1.15.1`. According to the official changelog (https://cffi.readthedocs.io/en/latest/whatsnew.html#v1-16-0rc1) support for python 3.12 was added in `1.16.0rc1`.

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -18,7 +18,6 @@ runs:
       run: |
         sudo apt-get update
         sudo apt-get install language-pack-en language-pack-de
-        sudo apt-get purge libffi-dev
       shell: bash
   
     - name: Setup Python
@@ -63,15 +62,26 @@ runs:
     - name: Install Python dependencies
       run: |
         python3 -V
+
+        # inspect pip cache
         python3 -m pip freeze | sort
         python3 -m pip cache info || true
         python3 -m pip cache list || true
-        python3 -m pip install --upgrade --force pip wheel
+
+        # remove cached built whl files
+        rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
+
+        # install dependencies
         python3 -m pip install --force -r python/requirements.txt
         python3 -m pip install --force -r python/test/requirements.txt -c python/test/constraints.txt
+
+        # inspect pip cache
         python3 -m pip freeze | sort
         python3 -m pip cache info || true
         python3 -m pip cache list || true
+
+        # assert no whl files have been built
+        python3 -m pip cache info | grep "Locally built wheels size:" | (! grep -v "Locally built wheels size: 0")
       shell: bash
   
     - name: Update expectation files

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -72,7 +72,8 @@ runs:
         rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
 
         # install dependencies
-        python3 -m pip install --force -r python/requirements.txt
+        python_minor_version="$(python3 -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"
+        python3 -m pip install --force -r python/requirements-$python_minor_version.txt
         python3 -m pip install --force -r python/test/requirements.txt -c python/test/constraints.txt
 
         # inspect pip cache

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -70,7 +70,7 @@ runs:
         python3 -m pip cache list || true
 
         # remove cached built whl files
-        #rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
+        rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
 
         # install dependencies
         python_minor_version="$(python3 -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -83,7 +83,7 @@ runs:
         python3 -m pip cache list || true
 
         # assert no whl files have been built
-        if [ "$(python3 -m pip cache info | grep "Number of locally built wheels:")" != "Number of locally built wheels: 0" ]
+        if [ python3 -m pip cache info ] && [ "$(python3 -m pip cache info | grep "Number of locally built wheels:")" != "Number of locally built wheels: 0" ]
         then
           echo "Dependency whl files have been built"
           exit 1

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -69,7 +69,8 @@ runs:
         python3 -m pip cache list || true
 
         # remove cached built whl files
-        rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
+        python3 -m pip cache purge || true
+        python3 -m pip cache list || true
 
         # install dependencies
         python_minor_version="$(python3 -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -62,7 +62,6 @@ runs:
     - name: Install Python dependencies
       run: |
         python3 -V
-        pip install --upgrade --force pip==24.0.0
 
         # inspect pip cache
         python3 -m pip freeze | sort

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -82,7 +82,11 @@ runs:
         python3 -m pip cache list || true
 
         # assert no whl files have been built
-        python3 -m pip cache info | grep "Locally built wheels size:" | (! grep -v "Locally built wheels size: 0")
+        if [ "$(python3 -m pip cache info | grep "Number of locally built wheels:")" != "Number of locally built wheels: 0" ]
+        then
+          echo "Dependency whl files have been built"
+          exit 1
+        fi
       shell: bash
   
     - name: Update expectation files

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -62,6 +62,7 @@ runs:
     - name: Install Python dependencies
       run: |
         python3 -V
+        pip install --upgrade --force pip==24.0.0
 
         # inspect pip cache
         python3 -m pip freeze | sort

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -69,7 +69,7 @@ runs:
         python3 -m pip cache list || true
 
         # remove cached built whl files
-        python3 -m pip cache purge || true
+        rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location: " | cut -d ":" -f 2- | cut -d " " -f 2-)"
         python3 -m pip cache list || true
 
         # install dependencies

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -18,6 +18,7 @@ runs:
       run: |
         sudo apt-get update
         sudo apt-get install language-pack-en language-pack-de
+        sudo apt-get purge libffi-dev
       shell: bash
   
     - name: Setup Python

--- a/.github/actions/test/action.yml
+++ b/.github/actions/test/action.yml
@@ -69,7 +69,7 @@ runs:
         python3 -m pip cache list || true
 
         # remove cached built whl files
-        rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
+        #rm -rf "$(python3 -m pip cache info | grep "Locally built wheels location:" | sed -e "s/.*: *//")"
 
         # install dependencies
         python_minor_version="$(python3 -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"

--- a/.github/upgrade-pip-packages.sh
+++ b/.github/upgrade-pip-packages.sh
@@ -2,12 +2,13 @@
 set -euo pipefail
 
 base="$(dirname "$0")"
+python_minor_version="$(python3 -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"
 
-pip install --upgrade --force pip==22.0.0
-pip install --upgrade --upgrade-strategy eager -r "$base/../python/requirements-direct.txt"
+pip install --upgrade --force pip==24.0.0
+pip install --upgrade --upgrade-strategy eager -r "$base/../python/requirements.txt"
 
 pip install pipdeptree
-pipdeptree --packages="$(sed -e "s/;.*//" -e "s/=.*//g" "$base/../python/requirements-direct.txt" | paste -s -d ,)" --freeze > "$base/../python/requirements.txt"
+pipdeptree --packages="$(sed -e "s/;.*//" -e "s/=.*//g" "$base/../python/requirements.txt" | paste -s -d ,)" --freeze > "$base/../python/requirements-$python_minor_version.txt"
 
-git diff "$base/../python/requirements.txt"
+git diff "$base/../python/requirements-$python_minor_version.txt"
 

--- a/.github/workflows/ci-cd.yml
+++ b/.github/workflows/ci-cd.yml
@@ -22,8 +22,8 @@ jobs:
         uses: actions/checkout@v4
       - name: Check requirements.txt
         run: |
-          (diff -w <(grep -v "python_version >  '3.7'" python/requirements.txt | sed -e "s/;.*//") python/requirements-3.7.txt) | (! grep -e "^<")
-          (diff -w <(grep -v "python_version <= '3.7'" python/requirements.txt | sed -e "s/;.*//") python/requirements-post-3.7.txt) | (! grep -e "^<")
+          (diff -w <(grep -v "python_version >  '3.7'" python/requirements.txt | sed -e "s/;.*//") python/requirements-3.7.txt || true) | (! grep -e "^<")
+          (diff -w <(grep -v "python_version <= '3.7'" python/requirements.txt | sed -e "s/;.*//") python/requirements-post-3.7.txt || true) | (! grep -e "^<")
         shell: bash
       - name: Check for dependency updates
         continue-on-error: true

--- a/.github/workflows/ci-cd.yml
+++ b/.github/workflows/ci-cd.yml
@@ -27,7 +27,7 @@ jobs:
         shell: bash
       - name: Check for dependency updates
         continue-on-error: true
-        run:
+        run: |
           pip install tox
           tox
           git diff --exit-code

--- a/.github/workflows/ci-cd.yml
+++ b/.github/workflows/ci-cd.yml
@@ -20,14 +20,17 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v4
-      - name: Check requirements.txt against requirements-direct.txt
+      - name: Check requirements.txt
         run: |
-          (diff -w python/requirements-direct.txt python/requirements.txt || true) | (! grep -e "^<")
+          (diff -w <(grep -v "python_version >  '3.7'" python/requirements.txt | sed -e "s/;.*//") python/requirements-3.7.txt) | (! grep -e "^<")
+          (diff -w <(grep -v "python_version <= '3.7'" python/requirements.txt | sed -e "s/;.*//") python/requirements-post-3.7.txt) | (! grep -e "^<")
         shell: bash
       - name: Check for dependency updates
         continue-on-error: true
         run:
-          .github/upgrade-pip-packages.sh
+          pip install tox
+          tox
+          git diff --exit-code
         shell: bash
 
   test-mac:

--- a/Dockerfile
+++ b/Dockerfile
@@ -14,7 +14,7 @@ RUN apk add --no-cache --upgrade expat libuuid
 COPY python/requirements.txt /action/
 RUN apk add --no-cache build-base libffi-dev; \
     pip install --upgrade --force --no-cache-dir pip && \
-    pip install --upgrade --force --no-cache-dir -r /action/requirements.txt; \
+    pip install --upgrade --force --no-cache-dir -r /action/requirements-post-3.7.txt; \
     apk del build-base libffi-dev
 
 COPY python/publish /action/publish

--- a/Dockerfile
+++ b/Dockerfile
@@ -11,10 +11,10 @@ LABEL com.github.actions.color="green"
 
 RUN apk add --no-cache --upgrade expat libuuid
 
-COPY python/requirements.txt /action/
+COPY python/requirements-post-3.7.txt /action/requirements.txt
 RUN apk add --no-cache build-base libffi-dev; \
     pip install --upgrade --force --no-cache-dir pip && \
-    pip install --upgrade --force --no-cache-dir -r /action/requirements-post-3.7.txt; \
+    pip install --upgrade --force --no-cache-dir -r /action/requirements.txt; \
     apk del build-base libffi-dev
 
 COPY python/publish /action/publish

--- a/README.md
+++ b/README.md
@@ -854,7 +854,7 @@ Self-hosted runners may require setting up a Python environment first:
     python-version: 3.8
 ```
 
-Start-up of the action is faster with `virtualenv` or `venv`, as well as `wheel` packages are installed.
+Start-up of the action is faster with `virtualenv` or `venv` package installed.
 
 ## Running as a composite action
 

--- a/composite/action.yml
+++ b/composite/action.yml
@@ -181,7 +181,15 @@ runs:
 
         echo "Python that creates venv: $PYTHON_BIN"
         echo "PYTHON_BIN=$PYTHON_BIN" >> "$GITHUB_ENV"
-        echo "version=$("$PYTHON_BIN" -V)" >> "$GITHUB_OUTPUT"
+
+        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}"))"
+        if [[ "$PYTHON_VERSION" == "3.7" ]]
+        then
+          echo "DEPENDENCIES_VERSION=3.7" >> "$GITHUB_ENV"
+        else
+          echo "DEPENDENCIES_VERSION=post-3.7" >> "$GITHUB_ENV"
+        fi
+        echo "version=$PYTHON_VERSION" >> "$GITHUB_OUTPUT"
 
         echo '::endgroup::'
       shell: bash
@@ -263,9 +271,7 @@ runs:
         # Install Python dependencies
         echo '::group::Install Python dependencies'
 
-        # make sure wheel is installed, which improves installing our dependencies
-        "$PYTHON_VENV" -m pip install wheel
-        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements.txt"
+        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements-$DEPENDENCIES_VERSION.txt"
 
         echo '::endgroup::'
       shell: bash

--- a/composite/action.yml
+++ b/composite/action.yml
@@ -182,7 +182,7 @@ runs:
         echo "Python that creates venv: $PYTHON_BIN"
         echo "PYTHON_BIN=$PYTHON_BIN" >> "$GITHUB_ENV"
 
-        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}"))"
+        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"
         if [[ "$PYTHON_VERSION" == "3.7" ]]
         then
           echo "DEPENDENCIES_VERSION=3.7" >> "$GITHUB_ENV"

--- a/composite/action.yml
+++ b/composite/action.yml
@@ -270,9 +270,7 @@ runs:
       run: |
         # Install Python dependencies
         echo '::group::Install Python dependencies'
-
         "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements-$DEPENDENCIES_VERSION.txt"
-
         echo '::endgroup::'
       shell: bash
 

--- a/composite/action.yml
+++ b/composite/action.yml
@@ -218,7 +218,7 @@ runs:
       continue-on-error: true
       with:
         path: ${{ steps.os.outputs.pip-cache }}
-        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-70a313922fdbeb7398ec60313d908b11
+        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-e594996205319a7990b3a4ec677d10a3
 
     - name: Create virtualenv
       id: venv

--- a/linux/action.yml
+++ b/linux/action.yml
@@ -176,7 +176,7 @@ runs:
         echo "Python that creates venv: $PYTHON_BIN"
         echo "PYTHON_BIN=$PYTHON_BIN" >> "$GITHUB_ENV"
 
-        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}"))"
+        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"
         if [[ "$PYTHON_VERSION" == "3.7" ]]
         then
           echo "DEPENDENCIES_VERSION=3.7" >> "$GITHUB_ENV"

--- a/linux/action.yml
+++ b/linux/action.yml
@@ -235,11 +235,7 @@ runs:
       run: |
         # Install Python dependencies
         echo '::group::Install Python dependencies'
-
-        # make sure wheel is installed, which improves installing our dependencies
-        "$PYTHON_VENV" -m pip install wheel
-        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements.txt"
-
+        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements-$DEPENDENCIES_VERSION.txt"
         echo '::endgroup::'
       shell: bash
 

--- a/linux/action.yml
+++ b/linux/action.yml
@@ -194,7 +194,7 @@ runs:
       continue-on-error: true
       with:
         path: '~/.cache/pip'
-        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-70a313922fdbeb7398ec60313d908b11
+        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-e594996205319a7990b3a4ec677d10a3
 
     - name: Create virtualenv
       id: venv

--- a/linux/action.yml
+++ b/linux/action.yml
@@ -175,7 +175,15 @@ runs:
 
         echo "Python that creates venv: $PYTHON_BIN"
         echo "PYTHON_BIN=$PYTHON_BIN" >> "$GITHUB_ENV"
-        echo "version=$("$PYTHON_BIN" -V)" >> "$GITHUB_OUTPUT"
+
+        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}"))"
+        if [[ "$PYTHON_VERSION" == "3.7" ]]
+        then
+          echo "DEPENDENCIES_VERSION=3.7" >> "$GITHUB_ENV"
+        else
+          echo "DEPENDENCIES_VERSION=post-3.7" >> "$GITHUB_ENV"
+        fi
+        echo "version=$PYTHON_VERSION" >> "$GITHUB_OUTPUT"
 
         echo '::endgroup::'
       shell: bash

--- a/macos/action.yml
+++ b/macos/action.yml
@@ -186,7 +186,7 @@ runs:
       continue-on-error: true
       with:
         path: '~/Library/Caches/pip'
-        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-70a313922fdbeb7398ec60313d908b11
+        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-e594996205319a7990b3a4ec677d10a3
 
     - name: Create virtualenv
       id: venv

--- a/macos/action.yml
+++ b/macos/action.yml
@@ -175,7 +175,15 @@ runs:
 
         echo "Python that creates venv: $PYTHON_BIN"
         echo "PYTHON_BIN=$PYTHON_BIN" >> "$GITHUB_ENV"
-        echo "version=$("$PYTHON_BIN" -V)" >> "$GITHUB_OUTPUT"
+
+        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"
+        if [[ "$PYTHON_VERSION" == "3.7" ]]
+        then
+          echo "DEPENDENCIES_VERSION=3.7" >> "$GITHUB_ENV"
+        else
+          echo "DEPENDENCIES_VERSION=post-3.7" >> "$GITHUB_ENV"
+        fi
+        echo "version=$PYTHON_VERSION" >> "$GITHUB_OUTPUT"
 
         echo '::endgroup::'
       shell: bash
@@ -227,11 +235,7 @@ runs:
       run: |
         # Install Python dependencies
         echo '::group::Install Python dependencies'
-
-        # make sure wheel is installed, which improves installing our dependencies
-        "$PYTHON_VENV" -m pip install wheel
-        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements.txt"
-
+        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../python/requirements-$DEPENDENCIES_VERSION.txt"
         echo '::endgroup::'
       shell: bash
 

--- a/python/requirements-3.10.txt
+++ b/python/requirements-3.10.txt
@@ -1,0 +1,18 @@
+humanize==4.9.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+  PyNaCl==1.5.0
+    cffi==1.16.0
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.2.1
+  typing_extensions==4.10.0
+  urllib3==2.2.1

--- a/python/requirements-3.11.txt
+++ b/python/requirements-3.11.txt
@@ -1,0 +1,18 @@
+humanize==4.9.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+  PyNaCl==1.5.0
+    cffi==1.16.0
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.2.1
+  typing_extensions==4.10.0
+  urllib3==2.2.1

--- a/python/requirements-3.12.txt
+++ b/python/requirements-3.12.txt
@@ -1,0 +1,18 @@
+humanize==4.9.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+  PyNaCl==1.5.0
+    cffi==1.16.0
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.2.1
+  typing_extensions==4.10.0
+  urllib3==2.2.1

--- a/python/requirements-3.7.txt
+++ b/python/requirements-3.7.txt
@@ -1,0 +1,22 @@
+humanize==4.6.0
+  importlib-metadata==6.7.0
+    typing_extensions==4.7.1
+    zipp==3.15.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+    typing_extensions==4.7.1
+  PyNaCl==1.5.0
+    cffi==1.15.1
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.0.7
+  typing_extensions==4.7.1
+  urllib3==2.0.7

--- a/python/requirements-3.8.txt
+++ b/python/requirements-3.8.txt
@@ -1,0 +1,18 @@
+humanize==4.9.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+  PyNaCl==1.5.0
+    cffi==1.16.0
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.2.1
+  typing_extensions==4.10.0
+  urllib3==2.2.1

--- a/python/requirements-3.9.txt
+++ b/python/requirements-3.9.txt
@@ -1,0 +1,18 @@
+humanize==4.9.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+  PyNaCl==1.5.0
+    cffi==1.16.0
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.2.1
+  typing_extensions==4.10.0
+  urllib3==2.2.1

--- a/python/requirements-direct.txt
+++ b/python/requirements-direct.txt
@@ -1,5 +1,0 @@
-humanize==3.14.0
-junitparser==3.1.0
-lxml==5.1.0
-psutil==5.9.7
-PyGithub==2.1.1

--- a/python/requirements-post-3.7.txt
+++ b/python/requirements-post-3.7.txt
@@ -1,0 +1,18 @@
+humanize==4.9.0
+junitparser==3.1.2
+lxml==5.1.0
+psutil==5.9.8
+PyGithub==2.2.0
+  Deprecated==1.2.14
+    wrapt==1.16.0
+  PyJWT==2.8.0
+  PyNaCl==1.5.0
+    cffi==1.16.0
+      pycparser==2.21
+  requests==2.31.0
+    certifi==2024.2.2
+    charset-normalizer==3.3.2
+    idna==3.6
+    urllib3==2.2.1
+  typing_extensions==4.10.0
+  urllib3==2.2.1

--- a/python/requirements.txt
+++ b/python/requirements.txt
@@ -1,25 +1,6 @@
-humanize==3.14.0
-junitparser==3.1.0
-  future==0.18.3
+humanize==4.6.0; python_version <= '3.7'
+humanize==4.9.0; python_version >  '3.7'
+junitparser==3.1.2
 lxml==5.1.0
-psutil==5.9.7
-PyGithub==2.1.1
-  Deprecated==1.2.14
-    wrapt==1.16.0
-  PyJWT==2.8.0
-  PyNaCl==1.5.0
-    # latest version that support Python 3.7
-    cffi==1.15.1
-      pycparser==2.21
-  python-dateutil==2.8.2
-    six==1.16.0
-  requests==2.31.0
-    certifi==2023.11.17
-    charset-normalizer==3.3.2
-    idna==3.6
-    # latest version that support Python 3.7
-    urllib3==2.0.7
-  # latest version that support Python 3.7
-  typing_extensions==4.7.1
-  # latest version that support Python 3.7
-  urllib3==2.0.7
+psutil==5.9.8
+PyGithub==2.2.0

--- a/python/test/constraints.txt
+++ b/python/test/constraints.txt
@@ -1,2 +1,0 @@
-# test_github.py fails with newer version
-Werkzeug<2.1.0

--- a/tox.ini
+++ b/tox.ini
@@ -1,0 +1,7 @@
+[tox]
+envlist = py{37,38,39,310,311,312}
+
+[testenv]
+commands = .github/upgrade-pip-packages.sh
+allowlist_externals = .github/upgrade-pip-packages.sh
+

--- a/windows/action.yml
+++ b/windows/action.yml
@@ -173,9 +173,15 @@ runs:
           Write-Output "Python that creates venv: $PYTHON_BIN"
           "PYTHON_BIN=$PYTHON_BIN" | Out-File -FilePath $env:GITHUB_ENV -Append
 
-          $VERSION = Invoke-Expression -Command "& '$PYTHON_BIN' -V"
-          Write-Output "Python version: $VERSION"
-          "version=$VERSION" | Out-File -FilePath $env:GITHUB_OUTPUT -Append
+          $PYTHON_VERSION = Invoke-Expression -Command "& '$PYTHON_BIN' -c 'import sys; print(f""{sys.version_info.major}.{sys.version_info.minor}"")'"
+          Write-Output "Python version: $PYTHON_VERSION"
+
+          if ( $PYTHON_VERSION -eq "3.7" ) {
+            "DEPENDENCIES_VERSION=3.7" | Out-File -FilePath $env:GITHUB_ENV -Append
+          } else {
+            "DEPENDENCIES_VERSION=post-3.7" | Out-File -FilePath $env:GITHUB_ENV -Append
+          }
+          "version=$PYTHON_VERSION" | Out-File -FilePath $env:GITHUB_OUTPUT -Append
         } finally {
           Write-Output "::endgroup::"
         }
@@ -229,9 +235,7 @@ runs:
         # Install Python dependencies
         Write-Output "::group::Install Python dependencies"
         try {
-          # make sure wheel is installed, which improves installing our dependencies
-          Invoke-Expression -Command "& '$env:PYTHON_VENV' -m pip install wheel"
-          Invoke-Expression -Command "& '$env:PYTHON_VENV' -m pip install -r '$env:GITHUB_ACTION_PATH\..\python\requirements.txt'"
+          Invoke-Expression -Command "& '$env:PYTHON_VENV' -m pip install -r '$env:GITHUB_ACTION_PATH\..\python\requirements-$env:DEPENDENCIES_VERSION.txt'"
         } finally {
           Write-Output "::endgroup::"
         }

--- a/windows/action.yml
+++ b/windows/action.yml
@@ -187,7 +187,7 @@ runs:
       continue-on-error: true
       with:
         path: '~\AppData\Local\pip\Cache'
-        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-70a313922fdbeb7398ec60313d908b11
+        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-e594996205319a7990b3a4ec677d10a3
 
     - name: Create virtualenv
       id: venv

--- a/windows/bash/action.yml
+++ b/windows/bash/action.yml
@@ -186,7 +186,7 @@ runs:
       continue-on-error: true
       with:
         path: '~\AppData\Local\pip\Cache'
-        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-70a313922fdbeb7398ec60313d908b11
+        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-e594996205319a7990b3a4ec677d10a3
 
     - name: Create virtualenv
       id: venv

--- a/windows/bash/action.yml
+++ b/windows/bash/action.yml
@@ -175,7 +175,15 @@ runs:
 
         echo "Python that creates venv: $PYTHON_BIN"
         echo "PYTHON_BIN=$PYTHON_BIN" >> "$GITHUB_ENV"
-        echo "version=$("$PYTHON_BIN" -V)" >> "$GITHUB_OUTPUT"
+
+        PYTHON_VERSION="$($PYTHON_BIN -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')"
+        if [[ "$PYTHON_VERSION" == "3.7" ]]
+        then
+          echo "DEPENDENCIES_VERSION=3.7" >> "$GITHUB_ENV"
+        else
+          echo "DEPENDENCIES_VERSION=post-3.7" >> "$GITHUB_ENV"
+        fi
+        echo "version=$PYTHON_VERSION" >> "$GITHUB_OUTPUT"
 
         echo '::endgroup::'
       shell: bash
@@ -231,11 +239,7 @@ runs:
       run: |
         # Install Python dependencies
         echo '::group::Install Python dependencies'
-
-        # make sure wheel is installed, which improves installing our dependencies
-        "$PYTHON_VENV" -m pip install wheel
-        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../../python/requirements.txt"
-
+        "$PYTHON_VENV" -m pip install -r "$GITHUB_ACTION_PATH/../../python/requirements-$DEPENDENCIES_VERSION.txt"
         echo '::endgroup::'
       shell: bash
 

